plugins/opensslcryptoplugin/evp/evp.cpp

```diff
@@ -848,28 +848,147 @@ int OpenSslEvp::sign(const EVP_MD *digestFunc,
 
-    r = EVP_DigestSignUpdate(mdctx, bytes, bytesCount);
+    *ctx = mdctx;
+    r = 1;
+    return r;
+
+    err_free_mdctx:
+    EVP_MD_CTX_destroy(mdctx);
+    *ctx = nullptr;
+    err_dontfree:
+    return r;
+}
+
+/*
+    int sign_session_update(EVP_MD_CTX *ctx,
+                            const void *bytes,
+                            size_t bytesCount);
+
+    Updates a signing session.
+
+    Arguments:
+    * ctx: the signing context to use, destroyed if an error occours
+    * bytes: data to sign
+    * bytesCount: the number of bytes in 'bytes'
+
+    Return value:
+    * 1 when the operation was successful.
+    * less than 0 when there was an error.
+*/
+
+int OpenSslEvp::sign_session_update(EVP_MD_CTX *mdctx,
+                        const void *bytes,
+                        size_t bytesCount)
+{
+    if (mdctx == nullptr) {
+        return -2;
+    }
+
+    int r = EVP_DigestSignUpdate(mdctx, bytes, bytesCount);
     OSSLEVP_HANDLE_ERR(r != 1, r = -1, "failed to update DigestSign", err_free_mdctx);
 
-    r = EVP_DigestSignFinal(mdctx, NULL, signatureLength);
+    r = 1;
+    return r;
+
+    err_free_mdctx:
+    EVP_MD_CTX_destroy(mdctx);
+    return r;
+}
+
+/*
+    int sign_session_finalize(EVP_MD_CTX *ctx,
+                              uint8_t **signature,
+                              size_t *signatureLength);
+
+    Finalizes the signing session, producing the signature.
+
+    Arguments:
+    * ctx: the signing context to use, destroyed if an error occours
+    * signature: where the generated signature will be stored, which will have to be freed using OPENSSL_free
+    * signatureLength: where the length of the generated signature will be stored
+
+    Return value:
+    * 1 when the operation was successful.
+    * less than 0 when there was an error.
+*/
+int OpenSslEvp::sign_session_finalize(EVP_MD_CTX *mdctx,
+                          uint8_t **signature,
+                          size_t *signatureLength)
+{
+    if (mdctx == nullptr) {
+        return -2;
+    }
+
+    int r = EVP_DigestSignFinal(mdctx, nullptr, signatureLength);
     OSSLEVP_HANDLE_ERR(r != 1, r = -1, "failed to finalize DigestSign (1st call)", err_free_mdctx);
 
     *signature = (uint8_t *) OPENSSL_malloc(*signatureLength);
-    OSSLEVP_HANDLE_ERR(*signature == NULL, r = -1, "failed to allocate memory for signature", err_free_mdctx);
+    OSSLEVP_HANDLE_ERR(*signature == nullptr, r = -1, "failed to allocate memory for signature", err_free_mdctx);
 
     r = EVP_DigestSignFinal(mdctx, *signature, signatureLength);
-    OSSLEVP_HANDLE_ERR(r != 1, r = -1; OPENSSL_free(*signature), "failed to finalize DigestSign (2nd call)", err_free_mdctx);
+    OSSLEVP_HANDLE_ERR(r != 1, r = -1; OPENSSL_free(*signature), "failed to finalize DigestSign (2nd call)", err_free_signature);
+
+    // Destroy the mdctx, then return the actual result
+    goto err_free_mdctx;
+    return r;
```

Comment:
isn't this return statement dead-code, as it will be bypassed by the goto above?

Reply:
Yes, it's very dead. Surprised the compiler didn't warn about it.